contracts/contracts/chains/ethereum/EthereumMessengers.sol

```diff
@@ -0,0 +1,112 @@
+
+contract EthereumL1Messenger is IMessenger, RestrictedCalls {
+    function callAllowed(address, address) external view returns (bool) {
+        // The call must be delivered by the Resolver
+        // Since we are on the same chain as the Resolver
+        // we should never see an execution of this function.
+        revert("Unexpected state.");
+    }
+
+    // Simply forward the message to the target on L1
+    /* solhint-disable avoid-tx-origin */
+    function sendMessage(address target, bytes calldata message)
+        external
+        restricted(block.chainid)
+    {
+        target.call(message);
+    }
+}
+
+// In order to keep naming consistent between our rollup specific contracts
+// we are using `EthereumL2Messenger` here as a name even if this contract
+// is supposed to be deployed on L1.
+contract EthereumL2Messenger is IMessenger, Ownable, RestrictedCalls {
+    enum MessageStatus {
+        UNDEFINED,
+        DEFINED,
+        RELAYED
+    }
+
+    Resolver public immutable resolver;
+
+    /// Maps message hashes with their status.
+    mapping(bytes32 => MessageStatus) public messageHashes;
+
+    constructor(address resolver_) {
+        resolver = Resolver(resolver_);
+    }
+
+    function callAllowed(address, address) external view returns (bool) {
+        // Since we are on the same chain as the FillManager & RequestManager
+        // we should never see an execution of this function.
+        revert("Unexpected state.");
+    }
+
+    // Stores the message hash for future use
+    function storeMessage(bytes calldata message) internal {
+        bytes32 messageHash = keccak256(message);
+
+        require(
+            messageHashes[messageHash] != MessageStatus.DEFINED,
+            "Message is already registered."
+        );
+        require(
+            messageHashes[messageHash] != MessageStatus.RELAYED,
+            "Message has already been relayed."
+        );
+
+        messageHashes[messageHash] = MessageStatus.DEFINED;
+    }
+
+    // Store message instead of forwarding it.
+    // Can be forwarded, if ever needed, by calling `relayMessage`.
+    //
+    // .. seealso:: :sol:func:`relayMessage`
+    function sendMessage(address, bytes calldata message)
+        external
+        restricted(block.chainid)
+    {
+        storeMessage(message);
+    }
+
+    function relayMessage(
+        bytes32 requestId,
+        bytes32 fillId,
+        uint256 sourceChainId,
+        address filler
+    ) external {
+        bytes32 messageHash = keccak256(
+            abi.encodeCall(
+                Resolver.resolve,
+                (requestId, fillId, block.chainid, sourceChainId, filler)
+            )
+        );
+
+        require(
+            messageHashes[messageHash] != MessageStatus.UNDEFINED,
+            "Message not yet registered."
+        );
+        require(
+            messageHashes[messageHash] != MessageStatus.RELAYED,
+            "Message has already been relayed."
+        );
```

Comment:
saw these two times maybe you can introduce a small check function for this.

Reply:
They are not the same.

1. In the first case we are checking that a message is not yet defined or relayed. If it is, we should not allow overwriting same messages.
2. In the second case we check if the message is not undefined or relayed. If it is undefined or already relayed then we should not try to relay that message.

I did it like this so we can provide different error messages for these cases for better understanding of whats going on.